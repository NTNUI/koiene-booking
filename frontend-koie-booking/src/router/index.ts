import Vue from 'vue';
import VueRouter from 'vue-router';
import store from '@/store';

Vue.use(VueRouter);

const routes = [
  {
    path: '/',
    name: 'AllKoier',
    component: () => import(/* webpackChunkName: "about" */ '../views/AllKoier.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/koie/:id',
    name: 'koie',
    component: () => import('../views/Koie.vue'),
    props: true,
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/booking/:id',
    name: 'booking',
    component: () => import('../views/Booking.vue'),
    meta: { requiresAuth: true, requiresAdmin: false }
  },
  {
    path: '/login',
    name: 'Login',
    component: () => import('../views/Login.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/admin',
    name: 'Admin',
    component: () => import('../views/Admin.vue'),
    meta: { requiresAuth: true, requiresAdmin: true }
  }
];

const router = new VueRouter({
  routes
});

export function beforeEach(to: any, from: any, next: any) {
  const isLoggedIn = store.getters['auth/isLoggedIn'];
  const isAdmin = store.getters['auth/isAdmin'];
  const requiresAuth = to.matched.some((record: any) => record.meta.requiresAuth);
  const requiresAdmin = to.matched.some((record: any) => record.meta.requiresAdmin);

  if (requiresAdmin && !isAdmin) next({ name: 'Login', query: { redirect: to.path } });
  else if (requiresAuth && !isLoggedIn) next({ name: 'Login', query: { redirect: to.path } });
  else if (to.name === 'Login' && isLoggedIn) next('/');
  else next();
}

router.beforeEach((to, from, next) => beforeEach(to, from, next));

export default router;
export { routes };
