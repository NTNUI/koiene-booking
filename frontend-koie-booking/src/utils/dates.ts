import dayjs from 'dayjs';
require('dayjs/locale/nb');
dayjs.locale('nb');

export function getDateString(baseDate?: string, addDays?: number): string {
  let resultString: string;
  if (baseDate && addDays) {
    resultString = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else if (baseDate) {
    resultString = new Date(baseDate).toISOString().substr(0, 10);
  } else if (addDays) {
    resultString = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else {
    resultString = new Date().toISOString().substr(0, 10);
  }
  return resultString;
}

export function getDate(baseDate?: string, addDays?: number): Date {
  let resultDate: Date;
  if (baseDate && addDays) {
    resultDate = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000);
  } else if (baseDate) {
    resultDate = new Date(baseDate);
  } else if (addDays) {
    resultDate = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000);
  } else {
    resultDate = new Date();
  }
  return resultDate;
}

export function addToDate(startDate: string, howMany: number, timeUnit: dayjs.OpUnitType): string {
  const date = dayjs(startDate + 'T00:00:00.000Z');
  return date.add(howMany, timeUnit).format('YYYY-MM-DD');
}

export function formatDate(dateISO: string, formatString: string): string {
  const date = dayjs(dateISO + 'T00:00:00.000Z');
  return date.format(formatString);
}
