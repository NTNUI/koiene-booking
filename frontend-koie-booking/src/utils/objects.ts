export const clone = (object: any) => Object.assign({}, object);

export const renameKey = (object: any, key: string, newKey: string): any => {
  const value = object[key];
  delete object[key];
  object[newKey] = value;
  return object;
};

export const sortObjectByKey = (unordered: any) => {
  let ordered: any;
  ordered = {};
  Object.keys(unordered)
    .sort()
    .forEach(function(key) {
      ordered[key] = unordered[key];
    });
  return ordered;
};
