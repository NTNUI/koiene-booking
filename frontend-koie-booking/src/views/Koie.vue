<template>
  <ErrorCard v-if="error" />
  <LoadingSpinner v-else-if="isLoading" />
  <v-container v-else :class="$style.container" class="global-max-width">
    <v-row>
      <v-col height="100%" cols="12" sm="8" class="pb-0 mb-0">
        <v-card :tile="true" :class="$style.imageCard" outlined class="global-card-outlined">
          <v-img height="100%" :src="imgUrl" />
        </v-card>
      </v-col>
      <v-col cols="12" xs="4" class="pt-0 d-xs-block d-sm-none">
        <v-card :class="$style.titleCard" :tile="true" outlined class="global-card-outlined-border-bottom">
          <v-card-title id="koieTitle" class="headline">{{ koieTitle }}</v-card-title>
        </v-card>
      </v-col>
      <v-col :class="$vuetify.breakpoint.xs ? 'mb-3' : 'mb-0'" class="pb-0">
        <v-card :tile="true" :class="$style.infoCard" outlined class="global-card-outlined">
          <v-card-title>{{ $t('koie.information') }}</v-card-title>
          <v-list flat dense inactive class="global-card-outlined px-4 pt-0 pb-4">
            <v-list-item-group color="primary">
              <v-list-item v-for="(item, i) in keyInfoItems" :key="i" :href="item.url" target="_blank" class="pa-0">
                <v-list-item-icon :class="$style.listItemIcon">
                  <v-icon v-text="item.icon"></v-icon>
                </v-list-item-icon>
                <v-list-item-content v-if="!item.isLink" :class="$style.listItemContent">
                  <p style="margin-bottom: 0">
                    {{ item.title }}: <span :class="$style.infoValue"> {{ item.text }}</span>
                  </p>
                </v-list-item-content>
                <v-list-item-content v-else :class="$style.listItemContent">
                  <p style="margin-bottom: 0">
                    {{ item.title }}:
                    <span :class="$style.infoValue">
                      <a :href="item.url" target="_blank">{{ item.text }}</a>
                    </span>
                  </p>
                </v-list-item-content>
              </v-list-item>
            </v-list-item-group>
          </v-list>
        </v-card>
      </v-col>
    </v-row>
    <!-- Hide the row below only on mobile -->
    <v-row class="d-none d-sm-flex">
      <v-col class="pt-0">
        <v-card :class="$style.titleCard" :tile="true" outlined class="global-card-outlined-border-bottom">
          <v-card-title id="koieTitle" class="display-1">{{ koieTitle }}</v-card-title>
        </v-card>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="12" sm="4" order-sm="last">
        <v-card :tile="true" :class="$style.calendarCard" outlined class="global-card-outlined mb-6">
          <v-card-title>{{ $t('koie.booking') }}</v-card-title>
          <Calendar :allkoier="false" :row="false" class="px-4" />
          <v-btn
            :disabled="disableBooking"
            :color="this.$scssVars.globalColorPrimary"
            :class="[$style.bookBtn]"
            class="mx-auto"
            raised="true"
            data-test="bookBtn"
            @click="goToBooking"
            >{{ $t('koie.bookingBtn') }}</v-btn
          >
          <v-alert
            v-if="disableBooking"
            v-model="disableBooking"
            transition="expand-transition"
            :class="$style.errorMsg"
            type="info"
            @click="error = false"
          >
            {{ noDatesAvailable ? $t('calendar.error_fullbooked') : $t('avalanche.error_msg') }}
          </v-alert>
        </v-card>
        <v-card
          v-if="this.$store.state.avalanche.koierToCheck.includes(koieTitle)"
          :tile="true"
          outlined
          class="global-card-outlined pb-4"
        >
          <v-card-title>{{ $t('koie.avalanche') }}</v-card-title>
          <v-card-subtitle
            ><a href="https://varsom.no/"> {{ $t('avalanche.reference') }} </a></v-card-subtitle
          >
          <AvalancheWarning :latitude="latitude" :longitude="longitude" class="px-4" />
        </v-card>
      </v-col>
      <v-col cols="12" sm="8" class="py-0" order-sm="first">
        <v-row>
          <v-col>
            <v-card :tile="true" outlined class="global-card-outlined">
              <v-card-title>{{ $t('koie.description') }}</v-card-title>
              <!-- eslint-disable-next-line -->
            <v-card-text v-html="koieDescription"></v-card-text>
              <v-btn
                :class="[$style.expandBtn]"
                class="mx-auto"
                :color="this.$scssVars.globalColorPrimary"
                @click="expandCard"
                >{{ expandDescriptionCard ? 'less' : 'more' }}</v-btn
              >
            </v-card>
          </v-col>
        </v-row>
        <v-row>
          <v-col>
            <v-card :tile="true" outlined class="global-card-outlined">
              <v-card-title>{{ $t('koie.parking') }}</v-card-title>
              <!-- eslint-disable-next-line -->
              <v-card-text v-html="parking"></v-card-text>
            </v-card>
          </v-col>
        </v-row>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import Calendar from '../components/Calendar.vue';
import ErrorCard from '../components/ErrorCard.vue';
import LoadingSpinner from '../components/LoadingSpinner.vue';
import AvalancheWarning from '../components/AvalancheWarning.vue';
import Vue from 'vue';
import { KoieData, KeyInfoItems } from '../types/koie';

export default Vue.extend({
  name: 'Koie',
  components: {
    Calendar,
    ErrorCard,
    LoadingSpinner,
    AvalancheWarning
  },
  data(): KoieData {
    return {
      koieDescription: '',
      expandDescriptionCard: false
    };
  },
  computed: {
    error(): boolean {
      return this.$store.state.koie.error;
    },
    dateFrom(): string {
      return this.$store.state.booking.dateFrom;
    },
    disableBooking(): boolean {
      return this.$store.state.avalanche.disableBooking;
    },
    noDatesAvailable(): boolean {
      return this.$store.state.booking.noDatesAvailable;
    },
    isLoading(): boolean {
      return this.$store.state.koie.isLoading;
    },
    koieTitle(): string {
      return this.$store.state.koie.koieData.name || this.$route.params.id.toLowerCase();
    },
    latitude(): string {
      return this.$store.state.koie.koieData.location.latitude;
    },
    longitude(): string {
      return this.$store.state.koie.koieData.location.longitude;
    },
    parking(): string {
      return this.$store.getters['koie/parking'];
    },
    imgUrl(): string {
      return '/koie_pictures/' + this.koieTitle.toLowerCase() + '1.jpg';
    },
    keyInfoItems(): Array<KeyInfoItems> {
      return [
        {
          title: this.$t('koie.keyInfo.beds'),
          text: this.$store.state.koie.koieData.number_of_beds,
          isLink: false,
          icon: 'mdi-bed'
        },
        {
          title: this.$t('koie.keyInfo.terrain'),
          text: this.$store.getters['koie/koieTerrain'],
          isLink: false,
          icon: 'mdi-terrain'
        },
        {
          title: this.$t('koie.keyInfo.altitude'),
          text: this.$store.state.koie.koieData.location.altitude + 'm',
          isLink: false,
          icon: 'mdi-pine-tree'
        },
        {
          title: this.$t('koie.keyInfo.position'),
          text: 'Google Maps',
          url: this.$store.getters['koie/position'],
          isLink: true,
          icon: 'mdi-crosshairs-gps'
        },
        {
          title: this.$t('koie.keyInfo.pdfMap'),
          text: 'See pdf',
          url: this.$store.state.koie.koieData.location.map_pdf,
          isLink: true,
          icon: 'mdi-map'
        },
        {
          title: this.$t('koie.keyInfo.physicalMap'),
          text: this.$store.state.koie.koieData.location.kartblad,
          url: 'http://org.ntnu.no/koiene/tips.php?p=3#1',
          isLink: true,
          icon: 'mdi-map'
        },
        {
          title: this.$t('koie.keyInfo.weather'),
          text: 'yr.no',
          url: this.$store.getters['koie/koieWeatherLink'],
          isLink: true,
          icon: 'mdi-weather-partly-cloudy'
        }
      ];
    }
  },
  created() {
    this.getKoieData(this.$route.params.id);
    window.addEventListener('resize', this.updateTextLength);
  },
  destroyed() {
    window.removeEventListener('resize', this.updateTextLength);
  },
  methods: {
    goToBooking() {
      this.$router.push(`/booking/${this.$route.params.id}`);
    },
    async getKoieData(koieName: string) {
      if (this.$store.state.koie.koieData.name.length === 0) {
        await this.$store.dispatch('koie/FETCH_DATA', koieName.toLowerCase());
      }
      this.setKoieDescription();
      this.updateTextLength();
    },
    //this method checks if the window is smaller than 600, and makes the text amount smaller if it is.
    updateTextLength() {
      if (window.innerWidth < 600) {
        this.koieDescription = this.koieDescription.slice(0, 300) + ' ...';
      } else {
        //if it is bigger then 600 it will set the text to inital length
        this.setKoieDescription();
      }
    },
    expandCard() {
      this.expandDescriptionCard = !this.expandDescriptionCard;
      this.expandDescriptionCard && this.setKoieDescription();
      !this.expandDescriptionCard && (this.koieDescription = this.koieDescription.slice(0, 300) + ' ...');
    },
    setKoieDescription() {
      let str = this.$store.getters['koie/koieDescription'];

      // Inserts an </b> after sentences that starts with a <b>-tag.
      str = str.replace(/<b>([^\.]+)\./g, function(x: string) {
        return x + '</b>';
      });
      this.koieDescription = str;
    }
  }
});
</script>

<style lang="scss" module>
.imageCard {
  border-radius: 4px 4px 4px 0px;
  height: 372px;
  @media only screen and (max-width: 960px) {
    height: 100%;
  }
}

.infoCard {
  height: 100%;
}

.calendarCard {
  display: flex;
  flex-direction: column;
}

.titleCard {
  border-radius: 0px 0px 4px 4px;
  width: 25%;
  min-width: 350px;
  @media only screen and (max-width: 600px) {
    min-width: 0;
    width: auto;
    max-width: 80%;
  }
}

.expandBtn {
  display: none;
  @media only screen and (max-width: 600px) {
    display: block;
    margin-bottom: 16px;
  }
}

.listItemIcon {
  margin: 6px 8px 4px 0 !important;
}

.listItemContent {
  padding: 4px 0 8px 0 !important;
}
.bookBtn {
  margin-top: 24px;
  margin-bottom: 16px;
}

.infoValue {
  display: inline;
  font-weight: 100;
}
.errorMsg {
  margin: 10px 16px 10px 16px;
}
</style>
