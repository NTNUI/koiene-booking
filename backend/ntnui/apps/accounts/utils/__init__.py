from accounts.utils.mail_utils import send_gsuite_confirmation_mail
