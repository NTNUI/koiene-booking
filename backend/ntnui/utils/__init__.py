from ntnui.utils.pagination import PageSize7, PageSize20, PageSize100
