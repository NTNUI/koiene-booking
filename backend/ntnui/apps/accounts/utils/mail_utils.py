from django.core.mail import send_mail
from django.template.loader import render_to_string

from ntnui.enums import MembershipType


def new_membership(membership):
    """ Sends confirmation email to users and guests who signs up for events and sub-events. """

    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [membership.member.email]
    subject = "Welcome"

    context = {"user": membership.member, "group": membership.group}

    # Mail body.
    plain_message = render_to_string("emails/new_membership.txt", context)
    html_message = render_to_string("emails/new_membership.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_gsuite_confirmation_mail(user, contact_email, password):
    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = "Opprettelse av ny NTNUI-bruker"

    context = {
        "first_name": user.first_name,
        "email": user.email,
        "contact_email": contact_email,
        "password": password,
    }

    # Mail body.
    plain_message = render_to_string("emails/new_gsuite_account.txt", context)
    html_message = render_to_string("emails/new_gsuite_account.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_welcome_mail_to_leader(user, group):
    """ Send a mail to a new leader """
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny leder i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_leader_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_leader_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_cashier(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny kasserer i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_cashier_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_cashier_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_board_member(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny tillitsvalgt i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_board_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_board_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail(membership):
    """ Delegate welcome mail to different board members """
    send_welcome_mail_to_board_member(user=membership.member, group=membership.group)
    if membership.type == MembershipType.leader:
        send_welcome_mail_to_leader(user=membership.member, group=membership.group)

    if membership.type == MembershipType.cashier:
        send_welcome_mail_to_cashier(user=membership.member, group=membership.group)
