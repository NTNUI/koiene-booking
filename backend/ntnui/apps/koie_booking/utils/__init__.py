from koie_booking.utils.mail_utils import send_confirmation_mail, send_koie_information_mail
from koie_booking.utils.date_utils import get_daterange
