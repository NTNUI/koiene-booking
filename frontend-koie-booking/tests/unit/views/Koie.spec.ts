import Vue from 'vue';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';
import flushPromises from 'flush-promises';

import mockAxios from 'jest-mock-axios';
import router from '@/router';
import { routes } from '@/router';

Vue.use(Vuetify);

// Components or views
import Koie from '../../../src/views/Koie.vue';
import { createWrapper } from '../utils';

describe('View Koie.vue', () => {
  it('Matches snapshot', async () => {
    const router = new VueRouter({ routes, mode: 'abstract' });
    router.push('/koie/flåkoia');
    await flushPromises();

    const wrapper = createWrapper(Koie, {
      router,
    });

    const responseObj = {
      data: { koie: { name: 'flåkoia', description: { description_eng: '' }, location: { terrain_eng: '' } } },
    };
    mockAxios.mockResponse(responseObj);
    await flushPromises();

    expect(wrapper).toMatchSnapshot();
  });

  it('BookBtn changes route', async () => {
    const router = new VueRouter({ routes, mode: 'abstract' });
    router.push('/koie/flåkoia');
    await flushPromises();

    const wrapper = createWrapper(Koie, {
      router,
    });

    const spy = jest.spyOn(wrapper.vm.$router, 'push');
    const responseObj = {
      data: { koie: { name: 'flåkoia', description: { description_eng: '' }, location: { terrain_eng: '' } } },
    };
    mockAxios.mockResponse(responseObj);
    await flushPromises();
    wrapper.find('[data-test="bookBtn"]').trigger('click');

    expect(spy).toHaveBeenCalledWith('/booking/flåkoia');
  });

  // This is a very important function of the app
  it('BookBtn disabled when disableBooking is true', async () => {
    const router = new VueRouter({ routes, mode: 'abstract' });
    router.push('/koie/flåkoia');
    await flushPromises();

    const wrapper = createWrapper(Koie, {
      router,
      computed: {
        disableBooking() {
          return true;
        },
      },
    });

    const responseObj = {
      data: { koie: { name: 'flåkoia', description: { description_eng: '' }, location: { terrain_eng: '' } } },
    };
    mockAxios.mockResponse(responseObj);
    await flushPromises();

    const bookBtn = wrapper.find('[data-test="bookBtn"]');

    expect(bookBtn.props().disabled).toBe(true);
  });
});
