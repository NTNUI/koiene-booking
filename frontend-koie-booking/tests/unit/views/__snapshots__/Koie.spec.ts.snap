// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Koie.vue Matches snapshot 1`] = `
<div class="container global-max-width">
  <div class="row">
    <div height="100%" class="pb-0 mb-0 col-sm-8 col-12">
      <div class="global-card-outlined v-card v-card--outlined v-sheet v-sheet--tile theme--light imageCard">
        <div class="v-responsive v-image" style="height: 100%;">
          <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;" name="fade-transition" mode="in-out"></div>
          <div class="v-responsive__content"></div>
        </div>
      </div>
    </div>
    <div xs="4" class="pt-0 d-xs-block d-sm-none col col-12">
      <div class="global-card-outlined-border-bottom v-card v-card--outlined v-sheet v-sheet--tile theme--light titleCard">
        <div id="koieTitle" class="v-card__title headline">flåkoia</div>
      </div>
    </div>
    <div class="pb-0 col mb-0">
      <div class="global-card-outlined v-card v-card--outlined v-sheet v-sheet--tile theme--light infoCard">
        <div class="v-card__title">Key information</div>
        <div role="list" inactive="" class="v-list global-card-outlined px-4 pt-0 pb-4 v-sheet v-sheet--tile theme--light v-list--dense v-list--flat">
          <div role="listbox" class="v-item-group theme--light v-list-item-group primary--text">
            <div tabindex="0" target="_blank" role="listitem" aria-selected="false" class="pa-0 v-list-item v-list-item--link theme--light">
              <div class="v-list-item__icon listItemIcon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-bed theme--light"></i></div>
              <div class="v-list-item__content listItemContent">
                <p style="margin-bottom: 0px;">
                  Beds: <span class="infoValue"> </span></p>
              </div>
            </div>
            <div tabindex="0" target="_blank" role="listitem" aria-selected="false" class="pa-0 v-list-item v-list-item--link theme--light">
              <div class="v-list-item__icon listItemIcon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-terrain theme--light"></i></div>
              <div class="v-list-item__content listItemContent">
                <p style="margin-bottom: 0px;">
                  Terrain: <span class="infoValue"> </span></p>
              </div>
            </div>
            <div tabindex="0" target="_blank" role="listitem" aria-selected="false" class="pa-0 v-list-item v-list-item--link theme--light">
              <div class="v-list-item__icon listItemIcon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-pine-tree theme--light"></i></div>
              <div class="v-list-item__content listItemContent">
                <p style="margin-bottom: 0px;">
                  Altitude: <span class="infoValue"> undefinedm</span></p>
              </div>
            </div><a tabindex="0" href="https://www.google.com/maps/place/undefined,undefined" target="_blank" role="listitem" aria-selected="false" class="pa-0 v-list-item v-list-item--link theme--light">
              <div class="v-list-item__icon listItemIcon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-crosshairs-gps theme--light"></i></div>
              <div class="v-list-item__content listItemContent">
                <p style="margin-bottom: 0px;">
                  Position:
                  <span class="infoValue"><a href="https://www.google.com/maps/place/undefined,undefined" target="_blank">Google Maps</a></span>
                </p>
              </div>
            </a>
            <div tabindex="0" target="_blank" role="listitem" aria-selected="false" class="pa-0 v-list-item v-list-item--link theme--light">
              <div class="v-list-item__icon listItemIcon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-map theme--light"></i></div>
              <div class="v-list-item__content listItemContent">
                <p style="margin-bottom: 0px;">
                  Pdf Map:
                  <span class="infoValue"><a target="_blank">See pdf</a></span>
                </p>
              </div>
            </div><a tabindex="0" href="http://org.ntnu.no/koiene/tips.php?p=3#1" target="_blank" role="listitem" aria-selected="false" class="pa-0 v-list-item v-list-item--link theme--light">
              <div class="v-list-item__icon listItemIcon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-map theme--light"></i></div>
              <div class="v-list-item__content listItemContent">
                <p style="margin-bottom: 0px;">
                  Physical Map:
                  <span class="infoValue"><a href="http://org.ntnu.no/koiene/tips.php?p=3#1" target="_blank"></a></span>
                </p>
              </div>
            </a>
            <div tabindex="0" target="_blank" role="listitem" aria-selected="false" class="pa-0 v-list-item v-list-item--link theme--light">
              <div class="v-list-item__icon listItemIcon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-weather-partly-cloudy theme--light"></i></div>
              <div class="v-list-item__content listItemContent">
                <p style="margin-bottom: 0px;">
                  Weather forecast:
                  <span class="infoValue"><a target="_blank">yr.no</a></span>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="row d-none d-sm-flex">
    <div class="pt-0 col">
      <div class="global-card-outlined-border-bottom v-card v-card--outlined v-sheet v-sheet--tile theme--light titleCard">
        <div id="koieTitle" class="v-card__title display-1">flåkoia</div>
      </div>
    </div>
  </div>
  <div class="row">
    <div class="col-sm-4 order-sm-last col-12">
      <div class="global-card-outlined mb-6 v-card v-card--outlined v-sheet v-sheet--tile theme--light calendarCard">
        <div class="v-card__title">Booking</div>
        <div class="container px-4 vbeds column">
          <div class="v-input v-input--is-label-active v-input--is-dirty theme--dark v-text-field v-text-field--is-booted">
            <div class="v-input__prepend-outer">
              <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">event</i></div>
            </div>
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-35" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">Arrival</label><input id="input-35" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
          </div>
          <div class="v-menu">
            <!---->
          </div>
          <div class="v-menu">
            <!---->
          </div>
          <div class="v-input v-input--is-label-active v-input--is-dirty theme--dark v-text-field v-text-field--is-booted">
            <div class="v-input__prepend-outer">
              <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">event</i></div>
            </div>
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-40" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">Departure</label><input id="input-40" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
          </div>
          <div class="divBeds">
            <p vbeds="true" class="bedsText">
              Beds
            </p>
            <p class="availableBedsText">( Infinity available )</p>
            <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--dark v-text-field v-text-field--single-line v-text-field--is-booted v-text-field--enclosed v-text-field--outlined bedsInput">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend><span>​</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><input min="1" max="Infinity" vbeds="true" id="input-44" type="number"></div>
                </div>
              </div>
            </div>
          </div>
          <!---->
        </div> <button type="button" class="mx-auto v-btn v-btn--contained theme--light v-size--default bookBtn" raised="true" data-test="bookBtn"><span class="v-btn__content">Book koie</span></button>
        <!---->
      </div>
      <!---->
    </div>
    <div class="py-0 col-sm-8 order-sm-first col-12">
      <div class="row">
        <div class="col">
          <div class="global-card-outlined v-card v-card--outlined v-sheet v-sheet--tile theme--light">
            <div class="v-card__title">Description</div>
            <div class="v-card__text"></div> <button type="button" class="mx-auto v-btn v-btn--contained theme--light v-size--default expandBtn"><span class="v-btn__content">more</span></button>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col">
          <div class="global-card-outlined v-card v-card--outlined v-sheet v-sheet--tile theme--light">
            <div class="v-card__title">Parking</div>
            <div class="v-card__text"></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
